### ISSUE TITLE
Different Versions Of Package

### ISSUE THREAD

**Issue by josuematos** [rwynn/monstache#142]
When i tried use a plugin to use a termQuery in function Processor i needed of package github.com/olivere/elastic. After compile monstache and new plugin with the same version of go, i receive this error message: 

"ERROR 2018/12/13 18:23:25 Unable to load mapper plugin /etc/monstache/wifi.so: plugin.Open("/etc/monstache/wifi"): plugin was built with a different version of package github.com/olivere/elastic/uritemplates
INFO 2018/12/13 18:23:25 Shutting down with exit status 1 after panic."

OS: CentOS7 3.10.0-862.14.4.el7.x86_64
Go Version: go1.11.2

Steps:
1 - go build monstache.go;
2 - go get -u github.com/olivere/elastic;
3 - go build -buildmode=plugin -o /etc/monstache/wifi.so test.go;
4 - monstache -mapper-plugin-path /etc/monstache/wifi4.so

Contents of file test.go:
package main
import (
    "context"
    "reflect"
    "github.com/olivere/elastic"
    "github.com/rwynn/monstache/monstachemap"
)

type Wifi struct {
        Project    string `json:"project_id"`
        Zone string `json:"zone_id"`
        Local string `json:"localizacao_id"`
}

func Process(input*monstachemap.ProcessPluginInput) error {
    doc := input.Document
    client := input.ElasticClient
    indexName := "wifi4media"
    termQuery := elastic.NewTermQuery("mac", doc["ap"])
    res, err := client.Search(indexName).
                Index(indexName).
                Query(termQuery).
                Do(context.Background())
    if err != nil {
        return err
    }

    var l Wifi
    for _, item := range res.Each(reflect.TypeOf(l)) {
        l := item.(Wifi)
        doc["project_id"] = l.Project
        doc["zone_id"] = l.Zone
        doc["localizacao_id"] = l.Local
    }

   return nil
}

**Comment by rwynn** [comment]
Go plugins require that the source versions are exactly the same between the code loading the plugin and the plugin itself. 

Since monstache is now using go native packages, you should build your plugin with the same dependencies as monstache.  I think this can be accomplished by putting `go.mod` and `go.sum` files in your plugin source code directory when doing a build. I also believe all builds should be done outside your $GOPATH.  

To summarize, git clone monstache but do this outside your $GOPATH.  Then do a `go install` inside the root folder.  Then copy your plugin source code into the monstache folder (so it's next to go.mod and go.sum) and then finally `go build -buildmode=plugin -o /etc/monstache/wifi.so test.go`.

**Comment by rwynn** [comment]
When you add a bulk index request to the processor make sure that any `_id` field is deleted from the source doc.

```
bulk := input.ElasticBulkProcessor
req := elastic.NewBulkIndexRequest()
delete(doc, "_id") // cannot have an _id in the source document
req.Doc(doc)
// the id goes on the req
req.Id("123")
bulk.Add(req)
```

**Comment by josuematos** [comment]
@rwynn thx for your quickly answer. I made all modifications in my code and the first issue is solved but now i am with another problem in bulkprocessor. I have the mongo doc and the elastic index, both are correctly, when i try add a new bulk a receive this error:
"ERROR 2018/12/14 11:53:56 elastic: bulk processor "monstache" was unable to perform work: elastic: Error 400 (Bad Request): Validation Failed: 1: index is missing;2: type is missing;"

This is my code:
"func Map(input *monstachemap.MapperPluginInput) (output *monstachemap.MapperPluginOutput, err error) {
    //output = &monstachemap.MapperPluginOutput{Skip: true}
    return
}

func Process(input*monstachemap.ProcessPluginInput) error {
    doc := input.Document
    client := input.ElasticClient
    indexName := "wifi4media"
    termQuery := elastic.NewTermQuery("mac", doc["ap_mac"])
    res, err := client.Search(indexName).
                Index(indexName).
                Query(termQuery).
                Do(context.Background())
    if err != nil {
        return err
    }

    if res.Hits.TotalHits > 0 {
        fmt.Printf("Found a total of %d wifis\n", res.Hits.TotalHits)
        // Iterate through results
        for _, hit := range res.Hits.Hits {
            // hit.Index contains the name of the index
            // Deserialize hit.Source into a Tweet (could also be just a map[string]interface{}).
            var t Wifi
            err := json.Unmarshal(*hit.Source, &t)
            if err != nil {
                // Deserialization failed
            }
            fmt.Printf("Wifi by %i: %s\n", t.Zone, t.ZoneName)
            doc["project_id"] = t.Project
            doc["zone_id"] = t.Zone
            doc["localizacao_id"] = t.Local
            doc["local_map_id"] = t.LocalMap
            doc["zone_name"] = t.ZoneName
            fmt.Printf("Zone NAME %s\n", doc["zone_name"])
            id := input.Document["_id"].(bson.ObjectId).Hex()
            bulk := input.ElasticBulkProcessor
            req := elastic.NewBulkIndexRequest()
            delete(doc, "_id") // cannot have an _id in the source document
            req.Doc(doc)
            req.Id(id)
            bulk.Add(req)
        }
    } else {
        // No hits
        fmt.Print("Found no data\n")
    }

    return nil
}
"
NOTE: If i uncomment the line with output Map ( //output = &monstachemap.MapperPluginOutput{Skip: true}) i get this error:
"panic: assignment to entry in nil map

goroutine 96 [running]:
main.mapDataGolang(0xc0000a8000, 0xc0004221e0, 0x0, 0x0)
	/usr/share/monstache/monstache.go:812 +0x2b4
main.mapData(0xc0000a8000, 0xc000218000, 0xc0004221e0, 0xc00027df48, 0x6c9f9b)
	/usr/share/monstache/monstache.go:821 +0x44
main.doIndex(0xc000218000, 0xc0000a8000, 0xc00020a680, 0xc0002e41a0, 0xc0004221e0, 0x6e203d2079617272, 0x4173496576697461)
	/usr/share/monstache/monstache.go:2540 +0x43
main.main.func5(0xc0004d9950, 0xc0004ac960, 0xc000218000, 0xc0000a8000, 0xc00020a680, 0xc0002e41a0)
	/usr/share/monstache/monstache.go:3673 +0xc7
created by main.main
	/usr/share/monstache/monstache.go:3670 +0x116f"

I think that now i need only makes the bulk works with the transformation made in mongo doc processed.

Thx for your help!

**Comment by josuematos** [comment]
Here my monstache.conf.toml:
"mongo-url = "mongodb://mongo-node1:27017"
elasticsearch-urls = ["http://localhost:9200"]
direct-read-namespaces = ["farma-1_stat.stat_archive"]
gzip = true
stats = true
index-stats = true
elasticsearch-max-conns = 4
replay = false
resume = true
resume-write-unsafe = false
resume-name = "default"
verbose = true
exit-after-direct-reads = false

[[script]]
namespace = "farma-1_stat.stat_archive"
script = """
module.exports = function(doc, ns) {
    if (doc.ap_mac) {
        find_accesspoint = findOne({mac: doc.ap_mac}, {
          database: "farma-1",
          collection: "device"
        });
        doc.ap_name = find_accesspoint.name;
        doc.fk_join = {
           name: "doc_child",
            parent: doc.ap_mac
        };

        doc._meta_monstache = { routing: doc.ap_mac };
    }

    if (doc.site_id) {
        find_site = findOne({_id: doc.site_id}, {
          database: "farma-1",
          collection: "site"
        });
        doc.site_name = find_site.name;
        doc.site_desc = find_site.desc;
   }
   var custom_type = ns.split(".")[1];
   doc.type = custom_type;
   doc.project_id = 0;
   doc.zone_id_id = 0;
   doc.local_map_id = 0;
   doc.localizacao_id = 0;
   doc.zone_name = "";

    return doc;
}
"""
"

**Comment by josuematos** [comment]
And here the trace of bulk:
"TRACE 2018/12/14 12:11:15 POST /_bulk HTTP/1.1^M
Host: localhost:9200^M
User-Agent: elastic/6.2.14 (linux-amd64)^M
Content-Length: 435160^M
Accept: application/json^M
Content-Type: application/x-ndjson^M
Accept-Encoding: gzip^M
^M
{"index":{"_id":"5bd8343ee4b01b6efb658cc6"}}
{"ap_mac":"f0:9f:c2:7c:4c:01","assoc_time":1540894470,"bytes":3257997,"channel":48,"datetime":"2018-10-30T10:14:30Z","disassoc_time":1540895469,"duration":999,"hostname":"iPhone-de-Marta","ip":"192.168.126.17","is_guest":false,"is_wired":false,"local_map_id":0,"localizacao_id":1693,"mac":"d4:a3:3d:46:57:a0","name":null,"o":"session","oid":"59c53311e4b080f97c3c421a","project_id":49,"radio":"na","roam_count":2,"rx_bytes":477967,"site_id":"59c53311e4b080f97c3c421a","time":1540894470000,"tx_bytes":2780030,"user_id":"5bc62f59e4b0ceb52de87f15","zone_id":95,"zone_name":"Zona Atendimento"}"

**Comment by rwynn** [comment]
You would need to add the Index and Type of the document to the request before adding it. 

See here for an example.  

https://github.com/rwynn/monstache/blob/master/monstache.go#L2453

**Comment by rwynn** [comment]
Sorry for now you can use the go plugin or scripts but not both.

**Comment by rwynn** [comment]
You can accomplish what your script is doing by implementing the Map function is a Golang plugin. It has access to the mgo Session.

**Comment by josuematos** [comment]
Yes, i put an update in comment. Ok about the scripts, your daemon is the best ETL to works with elastic, in my opnion. The site needs more example and documentation, but teh daemon is awesome. Many, many thanks!

**Comment by rwynn** [comment]
Thanks!  If you have ideas about what would have been helpful for you to know please send a PR by going to https://rwynn.github.io/monstache-site/ and clicking Edit on Github.
